src/tss2-tcti/tcti-py.c

```diff
@@ -0,0 +1,526 @@
+        goto py_error;
+    }
+
+    unsigned long long magic = tmp_result ? PyLong_AsUnsignedLongLong(tmp_result) : TCTI_PY_MAGIC;
+    TSS2_TCTI_MAGIC (tcti_common) = magic;
+
+    rc = TSS2_RC_SUCCESS;
+
+py_error:
+    PyErr_Print();
+    SAFE_DECREF(tmp_result);
+    SAFE_DECREF(tmp_arg);
+
+    return rc;
+}
+
+TSS2_RC tcti_py_init_py_module (TSS2_TCTI_PY_CONTEXT *tcti_py) {
+
+    TSS2_RC rc = TSS2_TCTI_RC_NOT_IMPLEMENTED;
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_py_down_cast (tcti_py);
+
+    rc = get_magic(tcti_py);
+    if (rc) {
+        goto error;
+    }
+
+    rc = GET_TRANSMIT_FROM_PY(tcti_py);
+    if (rc) {
+        goto error;
+    }
+
+    rc = GET_RECIEVE_FROM_PY(tcti_py);
+    if (rc) {
+        goto error;
+    }
+
+    TSS2_TCTI_VERSION (tcti_common) = TCTI_PY_VERSION;
+    TSS2_TCTI_TRANSMIT (tcti_common) = tcti_py_transmit;
+    TSS2_TCTI_RECEIVE (tcti_common) = tcti_py_receive;
+    TSS2_TCTI_FINALIZE (tcti_common) = tcti_py_finalize;
+
+    /* no make sticky or get_poll_handles */
+    TSS2_TCTI_MAKE_STICKY (tcti_common) = tcti_make_sticky_not_implemented;
+    TSS2_TCTI_GET_POLL_HANDLES (tcti_common) = tcti_py_get_poll_handles_not_implemented;
+
+    tcti_common->state = TCTI_STATE_TRANSMIT;
+    tcti_common->locality = 0;
+
+    rc = TSS2_RC_SUCCESS;
+
+error:
+    return rc;
+
+}
+
+TSS2_RC
+Tss2_Tcti_Py_Init (TSS2_TCTI_CONTEXT *tcti_context, size_t *size,
+        const char *conf)
+{
+    TSS2_RC rc = TSS2_TCTI_RC_GENERAL_FAILURE;
+
+    TSS2_TCTI_PY_CONTEXT *tcti_py = tcti_py_context_cast(tcti_context);
+
+     char *conf_copy = NULL;
+    PyObject *py_module = NULL, *py_modname = NULL, *py_arg = NULL;
+
+    if (size == NULL) {
+        LOG_ERROR("size pointer must be valid");
+        return TSS2_TCTI_RC_BAD_VALUE;
+    }
+
+    if (!conf) {
+        LOG_ERROR("must have a module name to run in conf");
+        return TSS2_TCTI_RC_BAD_VALUE;
+    }
+
+    if (tcti_context == NULL) {
+        *size = sizeof (TSS2_TCTI_PY_CONTEXT);
+        return TSS2_RC_SUCCESS;
+    }
+
+    memset(tcti_context, 0, *size);
+
+    conf_copy = strdup(conf);
+    if (!conf_copy) {
+        return TSS2_TCTI_RC_MEMORY;
+    }
+
+    char *arg = NULL;
+    char *module = conf_copy;
+    char *sep = strchr(conf_copy, ':');
+    if (sep) {
+        *sep = '\0';
+        arg = &sep[1];
+    }
+
+    LOG_DEBUG ("Initializing Python TCTI with module: \"%s\" arg \"%s\"",
+            module, arg ? arg : "(null)");
+
+    /* Increment the reference count on PYTHON init */
+    pthread_mutex_lock(&_global.lock);
+    size_t new_cnt = 0;
+    bool overflow = __builtin_add_overflow(_global.refcnt, 1, &new_cnt);
+    if (!overflow) {
+        _global.refcnt = new_cnt;
+        if (new_cnt == 1) {
+            /* See: https://stackoverflow.com/questions/60719987/embedding-python-which-uses-numpy-in-c-doesnt-work-in-library-dynamically-loa */
+            _global.dlhandle = dlopen ("libpython3.so", RTLD_LAZY|RTLD_GLOBAL);
```

Comment:
Why do we not link `libpython3.so` as a shared library? That would lead to compile-time errors if the libpython dependency is not installed (instead of runtime errors). Also, it would simplify the code.

Reply:
We actually do link it, but if other bindings are in the link chain, like CFFI, it will resolve the symbols for that and leave some of the symbols unresolved, so we add GLOBAL and LAZY to force the resolution of undefined symbols at run time.

That comment above is what led me down this path, it seems to work as expected. I didn't find any other way to work around it.

Reply:
We actually do link it, but if other bindings are in the link chain, like CFFI, it will resolve the symbols for that and leave some of the symbols unresolved, so we add GLOBAL and LAZY to force the resolution of undefined symbols at run time.

That comment above is what led me down this path, it seems to work as expected. I didn't find any other way to work around it.

Reply:
Hm, I'm not sure I have fully understood the issue. You mean because the python module uses tctildr?
Another thought: we need to get the pkg-config library name (for me "libpython3.10.so") via define, I think. For me, I had to change the hardcoded shared library name.

Reply:
> Hm, I'm not sure I have fully understood the issue. You mean because the python module uses tctildr? Another 

No it has to do with symbol resolution and the dynamic linker loader, not anything tss2 specific.

Reply:
> we need to get the pkg-config library name (for me "libpython3.10.so") via define

Did that in the newest spin